src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -0,0 +1,368 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.Closeable;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.apache.cassandra.io.util.FileUtils;
+
+/**
+ * Modified from {@link org.apache.cassandra.index.sasi.utils.RangeIterator} to support:
+ * 1. no generic type to reduce allocation
+ * 2. CONCAT iterator type
+ */
+public abstract class RangeIterator extends AbstractIterator<PrimaryKey> implements Closeable
```

Comment:
Some general thoughts on naming, none of which absolutely need to be addressed right this minute...

As it stands `RangeIterator` iterates over a range of primary keys, and `KeyRangeIterator` iterates over a range of primary keys in memory. So you could say the former is a `KeyRangeIterator` and the latter an `InMemoryKeyRangeIterator`...or just `RangeIterator` and `InMemoryRangeIterator`. Open to other suggestions...

Reply:
I agree that naming here is not good and ideally this should be named `AbstractPrimaryKeyRangeIterator`. That just seems a bit of mouthful, particularly when we then end up with InMemoryPrimaryKeyRangeIterator`, `ConcatenatingPrimaryKeyRangeIterator`, `UnioningPrimaryKeyRangeIterator` and `IntersectingPrimaryKeyRangeIterator`. Do you think these names are to verbose? I feel that if we are going to rename them then renaming to `KeyRangeIterator` is just a halfway change.

WDYT?

Reply:
I'd be fine w/ just leaving the "Primary" part out, given that makes things a little less verbose...and still an improvement? (Could also keep things like `KeyRangeUnionIterator`, etc. if you think it reads better.) Almost anything would be an improvement over the `RangeIterator` / `KeyRangeIterator` situation.

Reply:
Going with:
- KeyRangeIterator
- InMemoryKeyRangeIterator
- KeyRangeConcatIterator
- KeyRangeIntersectionIterator
- KeyRangeUnionIterator